politeiawww/email.go

```diff
@@ -0,0 +1,482 @@
+package main
+
+import (
+	"bytes"
+	"net/url"
+	"text/template"
+
+	"github.com/dajohi/goemail"
+
+	"github.com/decred/politeia/politeiawww/api/v1"
+	"github.com/decred/politeia/politeiawww/database"
+)
+
+const (
+	fromAddress      = "noreply@decred.org"
+	politeiaMailName = "Politeia"
+)
+
+func createBody(tpl *template.Template, tplData interface{}) (string, error) {
+	var buf bytes.Buffer
+	err := tpl.Execute(&buf, tplData)
+	if err != nil {
+		return "", err
+	}
+
+	return buf.String(), nil
+}
+
+func (b *backend) createEmailLink(path, email, token string) (string, error) {
+	l, err := url.Parse(b.cfg.WebServerAddress + path)
+	if err != nil {
+		return "", err
+	}
+
+	q := l.Query()
+	if email != "" {
+		q.Set("email", email)
+	}
+	if token != "" {
+		q.Set("verificationtoken", token)
+	}
+	l.RawQuery = q.Encode()
+
+	return l.String(), nil
+}
+
+func (b *backend) sendEmail(
+	subject, body string,
+	addToAddressesFn func(*goemail.Message) error,
+) error {
+	msg := goemail.NewMessage(fromAddress, subject, body)
+	err := addToAddressesFn(msg)
+	if err != nil {
+		return err
+	}
+
+	msg.SetName(politeiaMailName)
+	return b.cfg.SMTP.Send(msg)
```

Comment:
Is this a blocking call? if it is, is that ok?

Reply:
Yes it's blocking, and I think it's okay because it happens within the thread which reads the channel for that event type.